google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/AppEngineAdminService.java

```diff
@@ -49,8 +73,54 @@ public static AppEngineAdminService getInstance() {
    * @throws IOException if there was a transient error connecting to the API
    */
   @Nullable
-  public abstract Application getApplicationForProjectId(@NotNull String projectId,
-      @NotNull Credential credential) throws IOException, GoogleApiException;
+  public Application getApplicationForProjectId(@NotNull String projectId,
+      @NotNull Credential credential) throws IOException, GoogleApiException {
+    try {
+      // Load from the cache if it exists, otherwise delegate to implementation classes.
+      return appEngineApplicationCache.get(projectId, () ->
+          fetchApplicationForProjectId(projectId, credential));
+
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof AppEngineApplicationNotFoundException) {
+        return null;
+      } else {
+        handleExecutionException(e);
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  /**
+   * Returns a list of all available App Engine Locations
+   *
+   * @param credential the authenticated user Credential
```

Comment:
lower case credential, or if you're referring to  the Credential type then {@link Credential}